hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBStoreAbstractIterator.java

```diff
@@ -94,16 +98,25 @@ private void setCurrentEntry() {
 
   @Override
   public final boolean hasNext() {
-    return rocksDBIterator.get().isValid() &&
-        (prefix == null || startsWithPrefix(key()));
+    try {
+      if (!seekDoneAtleastOnce) {
+        throw new NoSuchElementException("Cannot check iterator hasNext before seeking either to start of iter or " +
+            "seeking to a specific key");
+      }
+      return currentEntry != null &&
+          (prefix == null || startsWithPrefix(currentEntry.getKey()));
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
```

Comment:
We should introduce new exceptions.  Why the exceptions were not needed previously?

Reply:
key() call from iterator is what is inefficient. In the implementation of this class it would be called from the iterator over and over again which would involve a buffer copy. In byte array it would be direct buffer to indirect buffer copy and in codec buffer it would be direct buffer to direct buffer copy which is still inefficient and unnecessary.